x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/ingest/InferenceProcessor.java

```diff
@@ -205,19 +204,65 @@ public void accept(ClusterState state) {
 
             int count = 0;
             for (PipelineConfiguration configuration : ingestMetadata.getPipelines().values()) {
+                Map<String, Object> configMap = configuration.getConfigAsMap();
                 try {
-                    Pipeline pipeline = Pipeline.create(configuration.getId(),
-                        configuration.getConfigAsMap(),
-                        ingestService.getProcessorFactories(),
-                        ingestService.getScriptService());
-                    count += pipeline.getProcessors().stream().filter(processor -> processor instanceof InferenceProcessor).count();
+                    List<Map<String, Object>> processorConfigs = ConfigurationUtils.readList(null, null, configMap, PROCESSORS_KEY);
+                    for (Map<String, Object> processorConfigWithKey : processorConfigs) {
+                        for (Map.Entry<String, Object> entry : processorConfigWithKey.entrySet()) {
+                            count += numInferenceProcessors(entry.getKey(), entry.getValue());
+                        }
+                    }
+                // We cannot throw any exception here. It might break other pipelines.
                 } catch (Exception ex) {
-                    logger.warn(new ParameterizedMessage("failure parsing pipeline config [{}]", configuration.getId()), ex);
+                    logger.debug(
+                        () -> new ParameterizedMessage("failed gathering processors for pipeline [{}]", configuration.getId()),
+                        ex);
                 }
             }
             currentInferenceProcessors = count;
         }
 
+        @SuppressWarnings("unchecked")
+        static int numInferenceProcessors(String processorType, Object processorDefinition) {
+            return numInferenceProcessors(processorType, (Map<String, Object>)processorDefinition, 0);
+        }
+
+        @SuppressWarnings("unchecked")
+        static int numInferenceProcessors(String processorType, Map<String, Object> processorDefinition, int level) {
+            int count = 0;
+            // arbitrary, but we must limit this somehow
+            if (MAX_INFERENCE_PROCESSOR_SEARCH_RECURSIONS > 10) {
+                return count;
+            }
+            if (processorType == null || processorDefinition == null) {
+                return count;
+            }
+            if (TYPE.equals(processorType)) {
+                count++;
+            }
+            if (FOREACH_PROCESSOR_NAME.equals(processorType)) {
+                Map<String, Object> innerProcessor = (Map<String, Object>)processorDefinition.get("processor");
+                if (innerProcessor != null) {
+                    for (Map.Entry<String, Object> innerProcessorWithName : innerProcessor.entrySet()) {
```

Comment:
`foreach` can only have 1 processor so there is no need to iterate here.

Reply:
The iteration is just for simplicity. Otherwise I will have assert size is == 1 and then get first entry, etc. Iteration here is cleaner IMO.

Reply:
👍 

It threw me because it is called the foreach processor and it looks like a misunderstanding, perhaps leave a comment